NitroxServer/GameLogic/ScheduleKeeper.cs

```diff
@@ -0,0 +1,202 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using NitroxModel.DataStructures;
+using NitroxModel.Packets;
+using NitroxServer.GameLogic.Unlockables;
+
+namespace NitroxServer.GameLogic
+{
+    public class ScheduleKeeper
+    {
+        public Dictionary<string, FakeScheduledGoal> Schedule = new Dictionary<string, FakeScheduledGoal>();
+        private ThreadSafeCollection<FakeScheduledGoal> scheduledGoals;
+        private PDAStateData pdaStateData;
+        private StoryGoalData storyGoalData;
+        private EventTriggerer eventTriggerer;
+        private PlayerManager playerManager;
+     
+        public ScheduleKeeper(ThreadSafeCollection<FakeScheduledGoal> scheduledGoals, PDAStateData pdaStateData, StoryGoalData storyGoalData)
+        {
+            if (scheduledGoals != null)
+            {
+                // This one var is the reference to the original one
+                this.scheduledGoals = scheduledGoals;
+                // We still want to get a "replicated" list in memory
+                for (int i = scheduledGoals.Count - 1; i >= 0; i--)
+                {
+                    FakeScheduledGoal scheduledGoal = scheduledGoals[i];
+                    // In the unlikely case that there's a duplicated entry
+                    if (Schedule.TryGetValue(scheduledGoal.GoalKey, out FakeScheduledGoal alreadyInGoal))
+                    {
+                        // We remove the goal that's already in if it's planned for later than the first one
+                        if (GetLatestOfTwoGoals(scheduledGoal, alreadyInGoal).TimeExecute == alreadyInGoal.TimeExecute)
+                        {
+                            UnScheduleGoal(alreadyInGoal.GoalKey);
+                        }
+                        // Else we simply don't add it
+                        continue;
+                    }
+                    
+                    Schedule.Add(scheduledGoal.GoalKey, scheduledGoal);
+                }
+            }
+            this.pdaStateData = pdaStateData;
+            this.storyGoalData = storyGoalData;
```

Comment:
Reorder the code here so that the fields are initialized first and then any conditional logic afterwards. That way you can return early.

```suggestion
            this.pdaStateData = pdaStateData;
            this.storyGoalData = storyGoalData;

            if (scheduledGoals == null)
            {
                return;
            }
            // This one var is the reference to the original one
            this.scheduledGoals = scheduledGoals;
            // We still want to get a "replicated" list in memory
            for (int i = scheduledGoals.Count - 1; i >= 0; i--)
            {
                FakeScheduledGoal scheduledGoal = scheduledGoals[i];
                // In the unlikely case that there's a duplicated entry
                if (Schedule.TryGetValue(scheduledGoal.GoalKey, out FakeScheduledGoal alreadyInGoal))
                {
                    // We remove the goal that's already in if it's planned for later than the first one
                    if (GetLatestOfTwoGoals(scheduledGoal, alreadyInGoal).TimeExecute == alreadyInGoal.TimeExecute)
                    {
                        UnScheduleGoal(alreadyInGoal.GoalKey);
                    }
                    // Else we simply don't add it
                    continue;
                }
                    
                Schedule.Add(scheduledGoal.GoalKey, scheduledGoal);
            }
```